optaweb-employee-rostering-backend/src/main/java/org/optaweb/employeerostering/service/solver/EmployeeRosteringConstraintProvider.java

```diff
@@ -0,0 +1,332 @@
+import static org.optaweb.employeerostering.domain.tenant.RosterConstraintConfiguration.CONSTRAINT_NO_2_SHIFTS_WITHIN_10_HOURS_FROM_EACH_OTHER;
+import static org.optaweb.employeerostering.domain.tenant.RosterConstraintConfiguration.CONSTRAINT_REQUIRED_SKILL_FOR_A_SHIFT;
+import static org.optaweb.employeerostering.domain.tenant.RosterConstraintConfiguration.CONSTRAINT_UNAVAILABLE_TIME_SLOT_FOR_AN_EMPLOYEE;
+import static org.optaweb.employeerostering.domain.tenant.RosterConstraintConfiguration.CONSTRAINT_UNDESIRED_TIME_SLOT_FOR_AN_EMPLOYEE;
+import static org.optaweb.employeerostering.domain.tenant.RosterConstraintConfiguration.CONSTRAINT_WEEKLY_MINUTES_MUST_NOT_EXCEED_CONTRACT_MAXIMUM;
+import static org.optaweb.employeerostering.domain.tenant.RosterConstraintConfiguration.CONSTRAINT_YEARLY_MINUTES_MUST_NOT_EXCEED_CONTRACT_MAXIMUM;
+
+public final class EmployeeRosteringConstraintProvider implements ConstraintProvider {
+
+    @Override
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[]{
+                lowRiskEmployeeAssignedToCovidWard(constraintFactory),
+                moderateRiskEmployeeAssignedToCovidWard(constraintFactory),
+                highRiskEmployeeAssignedToCovidWard(constraintFactory),
+                extremeRiskEmployeeAssignedToCovidWard(constraintFactory),
+                inoculatedEmployeeOutsideCovidWard(constraintFactory),
+                uniformDistributionOfInoculatedHours(constraintFactory),
+                maximizeInoculatedHours(constraintFactory),
+                migrationBetweenCovidAndNonCovidWards(constraintFactory),
+                nonCovidShiftStartedLessThan8HoursAfterFinishingCovidShift(constraintFactory),
+                requiredSkillForShift(constraintFactory),
+                unavailableEmployeeTimeSlot(constraintFactory),
+                atMostOneShiftPerDayPerEmployee(constraintFactory),
+                noTwoShiftsWithin10HoursFromEachOther(constraintFactory),
+                dailyMinutesMustNotExceedContractMaximum(constraintFactory),
+                weeklyMinutesMustNotExceedContractMaximum(constraintFactory),
+                monthlyMinutesMustNotExceedContractMaximum(constraintFactory),
+                yearlyMinutesMustNotExceedContractMaximum(constraintFactory),
+                assignEveryShift(constraintFactory),
+                undesiredEmployeeTimeSlot(constraintFactory),
+                desiredEmployeeTimeSlot(constraintFactory),
+                employeeNotRotationEmployee(constraintFactory)
+        };
+    }
+
+    private Constraint riskEmployeeAssignedToCovidWard(ConstraintFactory constraintFactory, CovidRiskType riskType,
+            String constraintName) {
+        return constraintFactory.from(Shift.class)
+                .filter(shift -> shift.getEmployee() != null)
```

Comment:
Doesn't `from` already do this (v.s. `fromUnfilitered`)?

Reply:
Yes and no.

`employee` on `Shift` is nullable, and therefore even initialized `Shift`s may have the value `null`. This is correct.